File: ios/AiBook/Core/Model/ShelfBookSorter.swift
import Foundation

// MARK: - ShelfSortOption（与安卓排序选项对齐）

enum ShelfSortOption: String, CaseIterable {
    case recentRead
    case importedAt
    case title
    case favoriteFirst
}

// MARK: - ShelfBookSorter（与安卓 ShelfBookSorter.kt 对齐）

enum ShelfBookSorter {
    static func sort(_ books: [LocalBook], by option: ShelfSortOption) -> [LocalBook] {
        switch option {
        case .recentRead:
            return books.sorted { a, b in
                let dateA = a.lastReadAt ?? .distantPast
                let dateB = b.lastReadAt ?? .distantPast
                return dateA == dateB ? a.id < b.id : dateA > dateB
            }
        case .importedAt:
            return books.sorted {
                $0.importedAt == $1.importedAt ? $0.id < $1.id : $0.importedAt > $1.importedAt
            }
        case .title:
            return books.sorted { a, b in
                let comparison = a.title.localizedCaseInsensitiveCompare(b.title)
                return comparison == .orderedSame ? a.id < b.id : comparison == .orderedAscending
            }
        case .favoriteFirst:
            return books.sorted { a, b in
                if a.favorite != b.favorite { return a.favorite }
                let dateA = a.lastReadAt ?? a.importedAt
                let dateB = b.lastReadAt ?? b.importedAt
                return dateA == dateB ? a.id < b.id : dateA > dateB
            }
        }
    }
}

File: android/core/model/bin/main/com/aibook/android/core/model/ShelfBookSorter.kt
package com.aibook.android.core.model

enum class ShelfSortOption(val label: String) {
    RECENT_READ("最近阅读"),
    IMPORTED_AT("最近加入"),
    TITLE("书名 A-Z"),
    AUTHOR("作者 A-Z"),
    PROGRESS("阅读进度")
}

object ShelfBookSorter {
    fun sort(books: List<LocalBook>, option: ShelfSortOption): List<LocalBook> {
        return when (option) {
            ShelfSortOption.RECENT_READ -> books.sortedWith(
                compareByDescending<LocalBook> { it.lastReadAt ?: it.importedAt }
                    .thenBy { it.title.lowercase() }
            )
            ShelfSortOption.IMPORTED_AT -> books.sortedWith(
                compareByDescending<LocalBook> { it.importedAt }
                    .thenBy { it.title.lowercase() }
            )
            ShelfSortOption.TITLE -> books.sortedBy { it.title.lowercase() }
            ShelfSortOption.AUTHOR -> books.sortedWith(compareBy<LocalBook> { it.author.orEmpty().lowercase() }.thenBy { it.title.lowercase() })
            ShelfSortOption.PROGRESS -> books.sortedByDescending { it.progress.percent }
        }
    }
}
